js/src/utils.ts

```diff
@@ -0,0 +1,75 @@
+import { PublicKey } from '@solana/web3.js';
+import type BN from 'bn.js';
+import type { ProgramAddresses, Snapshot } from './types';
+
+/**
+ * Derives the addresses from seed and solido program
+ * @param programAddresses Addresses of the program deployment
+ * @param additionalSeedString Seed string
+ * @returns Program derived address
+ */
+export const findAuthorityProgramAddress = async (
+  programAddresses: ProgramAddresses,
+  additionalSeedString: string
+) => {
+  const bufferArray = [
+    programAddresses.solidoInstanceId.toBuffer(),
+    Buffer.from(additionalSeedString),
+  ];
+
+  return (
+    await PublicKey.findProgramAddress(
+      bufferArray,
+      programAddresses.solidoProgramId
+    )
+  )[0];
+};
+
+/**
+ * Derives the stake account address from seed and validator's vote account address
+ * @param solidoInstanceId Solido instance account address
+ * @param solidoProgramId Solido program address
+ * @param validatorVoteAccount Vote account for the validator
+ * @param seed Seed to derive the address
+ * @returns Stake account address
+ */
+export const calculateStakeAccountAddress = async (
+  solidoInstanceId: PublicKey,
+  solidoProgramId: PublicKey,
+  validatorVoteAccount: PublicKey,
+  seed: BN
+) => {
+  const bufferArray = [
+    solidoInstanceId.toBuffer(),
+    validatorVoteAccount.toBuffer(),
+    Buffer.from('validator_stake_account'),
+    Buffer.from(seed.toArray('le', 8)),
+  ];
+
+  const [stakeAccountAddress] = await PublicKey.findProgramAddress(
+    bufferArray,
+    solidoProgramId
+  );
+
+  return stakeAccountAddress;
+};
+
+/**
+ * Get the stake account that has the most amount of SOL staked(heaviest)
+ * @param snapshot Snapshot of the Solido stats
+ * @returns Heaviest stake account
+ */
+export const getHeaviestValidatorStakeAccount = (
+  snapshot: Snapshot
+): Snapshot['validatorsStakeAccounts'][0] => {
+  const sortedValidatorStakeAccounts = snapshot.validatorsStakeAccounts.sort(
```

Comment:
To get the validator with the most stake, we don’t need to sort (which has O(n log n) complexity, where n is the number of validators.) A single pass is sufficient. (Which has O(n) complexity.)

```typescript
let heaviestStakeAccount = snapshot.validatorsStakeAccounts[0];
for (stakeAccount in snapshot.validatorsStakeAccounts) {
    if (stakeAccount.balance > heaviestStakeAccount.balance) {
        heaviestStakeAccount = stakeAccount;
    }
}
return heaviestStakeAccount;
```
(Assuming we can overload `operator>` for `Lamports`.)